pkg/solana/logpoller/loader.go

```diff
@@ -279,3 +291,218 @@ func (c *EncodedLogCollector) loadSlotBlocksRange(ctx context.Context, start, en
+	p.engine.GoTick(services.NewTicker(time.Second), p.run)
+
+	return nil
+}
+
+func (p *orderedParser) close() error {
+	return nil
+}
+
+func (p *orderedParser) addToExpectations(evt ProgramEvent) error {
+	expectations, ok := p.expect[evt.SlotNumber]
+	if !ok {
+		return fmt.Errorf("%w: %d", errExpectationsNotSet, evt.SlotNumber)
+	}
+
+	evts, ok := p.actual[evt.SlotNumber]
+	if !ok {
+		return fmt.Errorf("%w: %d", errExpectationsNotSet, evt.SlotNumber)
+	}
+
+	p.actual[evt.SlotNumber] = append(evts, evt)
+
+	if expectations == len(evts)+1 {
+		p.setReady(evt.SlotNumber)
+	}
+
+	return nil
+}
+
+func (p *orderedParser) expectations(block uint64) (int, bool, error) {
+	expectations, ok := p.expect[block]
+	if !ok {
+		return 0, false, fmt.Errorf("%w: %d", errExpectationsNotSet, block)
+	}
+
+	evts, ok := p.actual[block]
+	if !ok {
+		return 0, false, fmt.Errorf("%w: %d", errExpectationsNotSet, block)
+	}
+
+	return expectations, expectations == len(evts), nil
+}
+
+func (p *orderedParser) clearExpectations(block uint64) {
+	delete(p.expect, block)
+	delete(p.actual, block)
+}
+
+func (p *orderedParser) setReady(slot uint64) {
+	p.ready = append(p.ready, slot)
+}
+
+func (p *orderedParser) run(_ context.Context) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	_ = p.sendReadySlots()
+}
+
+func (p *orderedParser) sendReadySlots() error {
+	rmvIdx := make([]int, 0)
+
+	// start at the lowest block and find ready blocks
+	for idx, block := range p.blocks {
+		// if no expectations are set, we are still waiting on information for the block.
+		// if expectations set and not met, we are still waiting on information for the block
+		// no other block data should be sent until this is resolved
+		exp, met, err := p.expectations(block)
+		if err != nil || !met {
+			break
+		}
+
+		// if expectations are 0 -> remove and continue
+		if exp == 0 {
+			p.clearExpectations(block)
+			rmvIdx = append(rmvIdx, idx)
+
+			continue
+		}
+
+		// if expectations set and met -> forward, remove, and continue
+
+		// to ensure ordered delivery, break from the loop if a ready block isn't found
+		// this function should be preceded by clearEmptyBlocks
+		rIdx, ok := getIdx(p.ready, block)
+		if !ok {
+			return nil
+		}
+
+		evts, ok := p.actual[block]
+		if !ok {
+			return errors.New("invalid state")
+		}
+
+		var errs error
+		for _, evt := range evts {
+			errs = errors.Join(errs, p.parser.Process(evt))
+		}
+
+		// need possible retry
+		if errs != nil {
+			return errs
+		}
+
+		p.ready = remove(p.ready, rIdx)
+		rmvIdx = append(rmvIdx, idx)
+
+		delete(p.expect, block)
+		delete(p.actual, block)
+	}
+
+	for count, idx := range rmvIdx {
+		p.blocks = remove(p.blocks, idx-count)
+	}
+
+	return nil
+}
+
+func getIdx[T any](slice []T, match T) (int, bool) {
+	for idx, value := range slice {
+		if reflect.DeepEqual(value, match) {
+			return idx, true
+		}
+	}
+
+	return -1, false
+}
+
+func remove[T any](slice []T, s int) []T {
+	return append(slice[:s], slice[s+1:]...)
+}
```

Comment:
Could use `slices.Delete`:
https://pkg.go.dev/slices#Delete

Reply:
Good to know. I have not used the `slices` package yet.